Only deep reset if default object is of the same type as current and is not null

alamar opened the pull request.

JerryShea commented on src/main/java/net/openhft/chronicle/wire/WireMarshaller.java:
```diff
@@ -486,13 +485,16 @@ public T defaultValue() {
     public void reset(T o) {
         try {
             for (FieldAccess field : fields) {
```
why don't we see if o is Resettable and then call o.reset() ?

JerryShea commented on src/main/java/net/openhft/chronicle/wire/WireMarshaller.java:
```diff
@@ -486,13 +485,16 @@ public T defaultValue() {
+                Class<?> resettableClass = field.resettableClass(defaultValue); // TODO FieldAccess should hold its defaultValue
+                if (resettableClass != null) {
+                    Object value = field.field.get(o); // TODO Replace with unsafe access
+                    if (value != null && resettableClass == value.getClass()) {
```
if value instanceof Resettable call reset() on it?

alamar replied:
Makes sense changing the check to `Resettable`.

alamar commented:
@JerryShea @nicktindall @yevgenp What do you think of this approach?

Notably, it does not seem that we can practically deep reset collections, as .iterator() is a garbage producing operation.

JerryShea commented:
> @JerryShea @nicktindall @yevgenp What do you think of this approach?
> 
> Notably, it does not seem that we can practically deep reset collections, as .iterator() is a garbage producing operation.

I don't think we should deep reset.

The user can override `reset()` for a DTO containing child DTOs, and [current behaviour](https://github.com/OpenHFT/Chronicle-Wire#a-note-on-wiresreset) is documented, even though it is confusing

nicktindall commented:
> @JerryShea @nicktindall @yevgenp What do you think of this approach?
> 
> Notably, it does not seem that we can practically deep reset collections, as .iterator() is a garbage producing operation.

I haven't really given it the review it deserves (it's hard to know what's going on without really checking the code out and  having a look). But I do think the described behaviour is problematic and we should strive to do better there.

One thing this kind of change would require IMO is a fairly extensive barrage of tests, and probably a description of how it works (to replace the paragraph explaining the buggy behaviour), it's a complex problem to solve and I'd be more comfortable if we agreed on some behaviour and then wrote up tests to confirm that the behaviour was implemented faithfully.

yevgenp commented on src/test/java/net/openhft/chronicle/wire/WireResetTest.java:
```diff
@@ -72,7 +70,7 @@ public void testDeepReset() {
         assertNull(event1.identifier.id);
         assertTrue(event1.identifier.permissions.isEmpty());
-        assertNull(event1.identifier.parent.id);
+        assertNull(event1.identifier.parent);
```
That's not a **deep** reset anymore. And is causing a GC.

alamar replied:
@yevgenp We need to return objects to their pristine (post-constructor) state. If post-constructed field value is null then we need to restore it to null also.
User can have different lifecycle for this field as to avoid GC.

alamar replied:
Note that it's not the only possibility, it's that we struggle to get team approval even from this toned down change. I encourage you to join the discussion on dev@.

alamar replied:
(note that in this case, resetting `parent` when child object is reset would definitely be a bug. Parent usually has multiple children and the subtree should not be all reset when only one child's `reset()` is called)

alamar committed:
Only deep reset if default object is of the same type as current and is not null

Change API to remove redundant defaults param
Improve README around reset()
Revert "Disable deep resetting for object fields #745"
This reverts commit 96bcb58e9ef4cca56845430488da47dbe780dabf.

alamar commented:
https://teamcity.chronicle.software/buildConfiguration/Chronicle_BuildAll_BumpedBranch/876320?buildTab=overview&hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true passed

maxim-ponomarev commented on README.adoc:
```diff
@@ -739,7 +739,7 @@ However, more complicated objects with object instance fields may experience und
 
 1. create a new instance of the object to be reset
-2. copy all fields from the new instance to the existing instance
+2. copy all fields by reference from the new instance to the existing instance (but see below)
```
Maybe we could provide a paragraph about "exception" right here (instead of making a footnote)?

maxim-ponomarev replied:
We can also mention, that resetting Resettable is a new behaviour, introduced in `*.25`.

maxim-ponomarev commented on README.adoc:
```diff
@@ -739,7 +739,7 @@ However, more complicated objects with object instance fields may experience und
 In order to `reset` a marshallable object, the process is as follows:
 
 1. create a new instance of the object to be reset
```
It sounds like we _**always**_ create a new instance of the to be reset, which is

1. Not true (AFAIK we use cached value for that)?
2. Doesn't really align with our general "garbage free" approach.

maxim-ponomarev commented on README.adoc:
```diff
@@ -781,6 +781,8 @@ true
 
+An exception to this is when the field value implements `Resettable` interface, when the default value of the field is not null, and the current field value is of exactly the same class as the default value. In this case, `value.reset()` will be called instead of overwriting it with default value by reference.
+
 In order to work around this, if necessary, the marshallable class can override `Marshallable.reset`:
```
Can we change an example below, do demonstrate, that now it's possible to change field class to `Resettable`, in order to:

1. Avoid having an undesired side-effect on reset.
2. Support a garbage-free behaviour out-of-the-box.

alamar committed:
Further improvements to documentation

nicktindall commented on src/main/java/net/openhft/chronicle/wire/WireMarshaller.java:
```diff
@@ -1113,15 +1153,27 @@ protected void setValue(Object o, @NotNull ValueIn read, boolean overwrite) thro
-        public void getAsBytes(Object o, @NotNull Bytes<?> bytes) throws IllegalAccessException {
-            bytes.writeUtf8(String.valueOf(field.get(o)));
+        protected void setDefaultValue(Object ignored, Object o) {
+            if (defaultValue != null && defaultValue instanceof Resettable && !(defaultValue instanceof DynamicEnum)) {
```
Should we have a specific `ResettableObjectFieldAccess`, to de-clutter the `ObjectFieldAccess`? instanceof strikes me as a smell here.

E.g. you could just have `ResettableObjectFieldAccess` which decorates `ObjectFieldAccess` to add the additional behaviour and is only used when `defaultValue` is not null, `Resettable` and not a `DynamicEnum`

I believe @JerryShea went on a rampage once because using instanceof had some performance implications.

nicktindall replied:
Also is it possible to deprecate the 2-param `setDefaultValue` seeing as `defaultValue` is ignored everywhere?

JerryShea replied:
Yes - the less instanceof the better. There can be some nasty and surprising performance implications

alamar replied:
`defaultValue` is used by default implementation of this method. I will keep it referenced.

nicktindall commented on src/main/java/net/openhft/chronicle/wire/WireMarshaller.java:
```diff
@@ -1622,6 +1699,22 @@ protected void setValue(Object o, ValueIn read, boolean overwrite) {
+            }
+
+            unsafePutObject(o, offset, defaultValue);
+        }
```
This seems pretty nasty, especially given collections are usually mutable, how does it compare to existing behaviour?

alamar replied:
Yes, that's what the default implementation of `copy()` does, which is called in this method by default.

alamar replied:
I've removed `setDefaultValue` for collections/maps since they were not doing anything that `copy()` won't do anyway.

nicktindall commented on src/main/java/net/openhft/chronicle/wire/WireMarshaller.java:
```diff
@@ -1796,6 +1910,22 @@ protected void setValue(Object o, ValueIn read, boolean overwrite) {
+                return;
+            }
+
+            unsafePutObject(o, offset, defaultValue);
```
Again this seems fairly dangerous. Interested to know how it compares to existing behaviour

alamar replied:
See above

nicktindall commented on src/main/java/net/openhft/chronicle/wire/WireMarshaller.java:
```diff
@@ -1113,15 +1153,27 @@ protected void setValue(Object o, @NotNull ValueIn read, boolean overwrite) thro
+                    return;
+
+                if (existingValue != null && existingValue.getClass() == defaultValue.getClass()) {
+                    ((Marshallable) existingValue).reset();
```
Could this end in an infinite loop with a self-referential DTO or e.g. a doubly-linked structure?

Not sure how you'd go about avoiding that in a garbage free way, but perhaps something to consider?

alamar replied:
Yes, it could; however, due to the safety of JVM, we'll just get a neat `StackOverflowError` out of it.

I don't think we could easily guard against this scenario. DTOs are supposed to be trees and not graphs, anyway, since it is assumed they are serialized eventually.

nicktindall reviewed (approved):
Looks mostly OK with some comments

alamar committed:
Undo setDefaultValue for collections
